tensorflowonspark/pipeline.py

```diff
@@ -0,0 +1,442 @@
+  args = None
+  argv = None
+  def merge_args_params(self):
+    local_args = copy.copy(self.args)                 # make a local copy of args
+    args_dict = vars(local_args)                      # get dictionary view
+    for p in self.params:
+      args_dict[p.name] = self.getOrDefault(p.name)   # update with params
+    return local_args
+
+class TFEstimator(Estimator, TFParams, HasInputMapping,
+                  HasClusterSize, HasNumPS, HasInputMode, HasProtocol, HasTensorboard, HasModelDir, HasExportDir, HasTFRecordDir,
+                  HasBatchSize, HasEpochs, HasReaders, HasSteps):
+  """Spark ML Pipeline Estimator which launches a TensorFlowOnSpark cluster for training"""
+
+  train_fn = None
+  export_fn = None
+
+  def __init__(self, train_fn, tf_args, tf_argv=None, export_fn=None):
+    """
+      :param train_fn: TensorFlow "main" function for training.
+      :param tf_args: Dictionary of arguments specific to TensorFlow "main" function.
+    """
+    super(TFEstimator, self).__init__()
+    self.train_fn = train_fn
+    self.export_fn = export_fn
+    self.args = Namespace(tf_args) if isinstance(tf_args, dict) else tf_args
+    self.argv = tf_argv
+    self._setDefault(input_mapping={},
+                    cluster_size=1,
+                    num_ps=0,
+                    input_mode=TFCluster.InputMode.SPARK,
+                    protocol='grpc',
+                    tensorboard=False,
+                    model_dir=None,
+                    export_dir=None,
+                    tfrecord_dir=None,
+                    batch_size=100,
+                    epochs=1,
+                    readers=1,
+                    steps=1000)
+
+  def _fit(self, dataset):
+    """Trains a TensorFlow model and returns a TFModel instance with params/args pointing to a checkpoint/saved_model on disk.
+    Note: for input_mode == InputMode.TENSORFLOW, the dataset input is ignored, and it's assumed that the TF train_fn is responsible for reading input data.
+    """
+    sc = SparkContext.getOrCreate()
+
+    logging.info("===== 1. train args: {0}".format(self.args))
+    logging.info("===== 2. train params: {0}".format(self._paramMap))
+    local_args = self.merge_args_params()
+    logging.info("===== 3. train args + params: {0}".format(local_args))
+
+    if local_args.input_mode == TFCluster.InputMode.TENSORFLOW:
+      if dfutil.isLoadedDF(dataset):
+        # if just a DataFrame loaded from tfrecords, just point to original source path
+        logging.info("Loaded DataFrame of TFRecord.")
+        local_args.tfrecord_dir = dfutil.loadedDF[dataset]
+      else:
+        # otherwise, save as tfrecords and point to save path
+        assert local_args.tfrecord_dir, "Please specify --tfrecord_dir to export DataFrame to TFRecord."
+        logging.info("Exporting DataFrame {} as TFRecord to: {}".format(dataset.dtypes, local_args.tfrecord_dir))
+        dfutil.saveAsTFRecords(dataset, local_args.tfrecord_dir)
```

Comment:
Should we only save the columns specified by inputMapping? That may reduce some cost.